Userland/Applications/Browser/BookmarksBarWidget.cpp

```diff
@@ -336,7 +347,7 @@ ErrorOr<void> BookmarksBarWidget::edit_bookmark(StringView url, PerformEditOn pe
         auto item_url = model()->index(item_index, 1).data().to_deprecated_string();
 
         if (item_url == url) {
-            auto values = BookmarkEditor::edit_bookmark(window(), item_title, item_url, perform_edit_on);
+            auto values = TRY(BookmarkEditor::edit_bookmark(window(), item_title, item_url, perform_edit_on));
```

Comment:
i feel that all the change needed was roughly
```cpp
auto values = BookmarkEditor::edit_bookmark(window(), item_title, item_url, perform_edit_on);

if (!values.is_empty()) {
    TRY(update_model(values, [item_index](auto& model, auto&& values) {
        return model.set(item_index, move(values));
    }));
    
    if (on_bookmark_change)
        on_bookmark_change();
}

return {};
```

Reply:
I agree with your point; I had a discussion on Discord on how to handle specific errors and using errnos seemed like the best way to go. I could also alternatively compare against an error string to see if we got a specific error as I believe that could stand the test of time.

Regarding your change, from the looks of it, it would solve the issue when editing a bookmark. However, if `BookmarksBarWidget::edit_bookmark()` were to return no error when a user cancels or aborts the editor dialog, `BookmarksBarWidget::add_bookmark()` would have no way of knowing that this had occurred so it will still add a bookmark against the user's wishes. If you feel like this coupling is an issue please let me know!

Reply:
Hi, I'm also on team "canceling is not an error". I'd rather see an `Optional<...>` here or another mechanism indicating that the dialog box did not result in a new value. Let's use `ErrorOr` only for situations where an error has actually occurred outside of the user's control.

Reply:
I agree. Big fan of using an `Optional` as the return type in `BookmarkEditor::edit_bookmark()` as it would more clearly communicate what's going on rather than having the caller check if the returned vector is empty.

> Regarding your change, from the looks of it, it would solve the issue when editing a bookmark. However, if `BookmarksBarWidget::edit_bookmark()` were to return no error when a user cancels or aborts the editor dialog, `BookmarksBarWidget::add_bookmark()` would have no way of knowing that this had occurred so it will still add a bookmark against the user's wishes. If you feel like this is an issue please let me know!

However, it seems that we will run into the same problem I mentioned earlier, since `BookmarksBarWidget::add_bookmark()` depends on the return value of `BookmarksBarWidget::edit_bookmark()`. If the editor dialog is canceled, `add_bookmark()` won't have a way of determining that and will keep the bookmark.

One solution I see would be to do something that's less DRY and duplicate some of `edit_bookmark()`'s code and add it to `add_bookmark()`. I'll go ahead and implement this now. If you see a better approach please let me know!

Reply:
I don't understand the problem. Return an `Optional<Vector<...>>`:
* When adding a new bookmark, when the `Optional` doesn't contain a value, do not add the bookmark.
* When editing a bookmark, when the `Optional` doesn't contain a value, do not update the bookmark.

Reply:
You might even choose to return an empty `Vector` instead of adding an `Optional` on top, I don't really care either way :^)

Reply:
> I don't understand the problem. Return an `Optional<Vector<...>>`:
> 
> * When adding a new bookmark, when the `Optional` doesn't contain a value, do not add the bookmark.
> * When editing a bookmark, when the `Optional` doesn't contain a value, do not update the bookmark.

Yes for editing a bookmark, we can implement `BookmarksWidget::edit_bookmark()` in such a way that it won't update the bookmark and not return an error if the `Optional` did not have a value. But for `add_bookmark()`, it initially adds a bookmark to the model before calling on `BookmarksBarWidget::edit_bookmark()`. This leads to the problem mentioned because `edit_bookmark()` won't return an error in the case where the user cancels the dialog, thereby leaving the new bookmark untouched when it should be deleted.

Reply:
> This leads to the problem mentioned because edit_bookmark() won't return an error in the case where the user cancels the dialog, thereby leaving the new bookmark untouched when it should be deleted.

Currently it detects an error and removes the bookmark. You can still do this if there is no value returned, right?

Reply:
Hmm for some reason I don't seem to understand. Yes, `add_bookmark()` currently with my changes would detect a specific error and remove a bookmark. `BookmarksBarWidget::edit_bookmark()` returns an `ErrorOr<void>`. From what you're suggesting, I'm interpreting it as I can still remove a bookmark if no error value was returned from `BookmarksBarWidget::edit_bookmark()`, which yes you can but we will lose that new bookmark we added.

As a side note thanks for helping me out thus far!

Reply:
To make it more concrete;

* `edit_bookmark()` should not return `ErrorOr`, just make it `Optional<Vector<JsonValue>>` or `Vector<JsonValue>` (and check whether the vector is empty)
* `add_bookmark()` invokes `edit_bookmark()`. If the `Optional` (or `Vector`) is empty, remove the bookmark you just added since the user canceled the dialog.

Reply:
Ah that is a bit confusing, yeah. I'd say let both `BookmarksBarWidget::add_bookmark()` and `BookmarksBarWidget::edit_bookmark()` call into `BookmarkEditor::edit_bookmark()`.

Reply:
Excellent! I'll make the necessary changes soon. Once again thanks for lending a hand!